File: library/src/main/java/name/zeno/android/third/baidu/PoiModel.java
package name.zeno.android.third.baidu;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;

import com.baidu.mapapi.search.core.PoiInfo;

import lombok.Data;

/**
 * 百度地图POI
 */
@Data @SuppressWarnings("WrongConstant")
public class PoiModel implements Parcelable
{
  public static final String COOKIE_LAST_POI = "bd_last_poi";
  private String name;
  private String address;
  private String city;
  @PoiTypeInt
  private int    type;
  private double latitude;
  private double longitude;
  private long   time;

  public PoiModel()
  {
    time = System.currentTimeMillis();
  }

  public PoiModel(@NonNull PoiInfo poiInfo)
  {
    this.name = poiInfo.name;
    this.address = poiInfo.address;
    this.city = poiInfo.city;
    if (poiInfo.type != null) {
      this.type = poiInfo.type.getInt();
    }

    if (poiInfo.location != null) {
      this.latitude = poiInfo.location.latitude;
      this.longitude = poiInfo.location.longitude;
    }
  }

  @IntDef({
      PoiTypeInt.POINT,
      PoiTypeInt.BUS_STATION,
      PoiTypeInt.BUS_LINE,
      PoiTypeInt.SUBWAY_STATION,
      PoiTypeInt.SUBWAY_LINE,
  })
  public @interface PoiTypeInt
  {
    int POINT          = 0;
    int BUS_STATION    = 1;
    int BUS_LINE       = 2;
    int SUBWAY_STATION = 3;
    int SUBWAY_LINE    = 4;
  }


  @Override public int describeContents() { return 0; }

  @Override public void writeToParcel(Parcel dest, int flags)
  {
    dest.writeString(this.name);
    dest.writeString(this.address);
    dest.writeString(this.city);
    dest.writeInt(this.type);
    dest.writeDouble(this.latitude);
    dest.writeDouble(this.longitude);
    dest.writeLong(this.time);
  }

  protected PoiModel(Parcel in)
  {
    this.name = in.readString();
    this.address = in.readString();
    this.city = in.readString();
    this.type = in.readInt();
    this.latitude = in.readDouble();
    this.longitude = in.readDouble();
    this.time = in.readLong();
  }

  public static final Parcelable.Creator<PoiModel> CREATOR = new Parcelable.Creator<PoiModel>()
  {
    @Override public PoiModel createFromParcel(Parcel source) {return new PoiModel(source);}

    @Override public PoiModel[] newArray(int size) {return new PoiModel[size];}
  };
}

File: library/src/main/java/name/zeno/android/third/baidu/PoiModel.kt
package name.zeno.android.third.baidu

import android.os.Parcel
import android.os.Parcelable
import android.support.annotation.IntDef

import com.baidu.mapapi.search.core.PoiInfo

/**
 * 百度地图POI
 */
class PoiModel(
    var name: String? = null,
    var address: String? = null,
    var city: String? = null,
    @PoiTypeInt
    var type: Int = POINT,
    var latitude: Double = 0.0,
    var longitude: Double = 0.0,
    var time: Long = System.currentTimeMillis()
) : Parcelable {
  @IntDef(POINT.toLong(), BUS_STATION.toLong(), BUS_LINE.toLong(), SUBWAY_STATION.toLong(), SUBWAY_LINE.toLong())
  annotation class PoiTypeInt()

  constructor(poiInfo: PoiInfo) : this(
      name = poiInfo.name,
      address = poiInfo.address,
      city = poiInfo.city,
      type = poiInfo.type?.int ?: POINT,
      latitude = poiInfo.location?.latitude ?: 0.0,
      longitude = poiInfo.location?.longitude ?: 0.0
  )

  constructor(source: Parcel) : this(
      source.readString(),
      source.readString(),
      source.readString(),
      source.readInt(),
      source.readDouble(),
      source.readDouble(),
      source.readLong()
  )

  override fun describeContents() = 0

  override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
    writeString(name)
    writeString(address)
    writeString(city)
    writeInt(type)
    writeDouble(latitude)
    writeDouble(longitude)
    writeLong(time)
  }

  companion object {
    const val COOKIE_LAST_POI = "bd_last_poi"

    const val POINT = 0

    const val BUS_STATION = 1

    const val BUS_LINE = 2

    const val SUBWAY_STATION = 3

    const val SUBWAY_LINE = 4

    @JvmField
    val CREATOR: Parcelable.Creator<PoiModel> = object : Parcelable.Creator<PoiModel> {
      override fun createFromParcel(source: Parcel): PoiModel = PoiModel(source)
      override fun newArray(size: Int): Array<PoiModel?> = arrayOfNulls(size)
    }
  }
}
